File: Ghost/Controllers/ViewControllers/AddBlog/AddBlogViewController.m
//
//  AddBlogViewController.m
//  Ghost
//
//  Created by Enric Enrich on 12/09/15.
//
//

#import <AFNetworking/AFNetworking.h>
#import <hpple/TFHpple.h>

#import "AddBlogViewController.h"

@interface AddBlogViewController ()

@property (nonatomic, strong) NSImageView *logoImageView;
@property (nonatomic, strong) NSTextField *blogUrlTextField;
@property (nonatomic, strong) NSButton *addBlogButton;

@end

@implementation AddBlogViewController

#pragma mark - Override

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    [self.view addSubview:self.logoImageView];
    [self.view addSubview:self.blogUrlTextField];
    [self.view addSubview:self.addBlogButton];
    
    NSDictionary *views = @{@"logoImageView": self.logoImageView,
                            @"blogUrlTextField": self.blogUrlTextField,
                            @"addBlogButton": self.addBlogButton};
    
    [self setUpConstraintsForViews:views];
}

#pragma mark - Getters

- (NSImageView *)logoImageView
{
    if (!_logoImageView) {
        _logoImageView = [NSImageView new];
        _logoImageView.image = [NSImage imageNamed:@"ghostLogoAddBlog"];
        _logoImageView.translatesAutoresizingMaskIntoConstraints = NO;
    }
    
    return _logoImageView;
}

- (NSTextField *)blogUrlTextField
{
    if (!_blogUrlTextField) {
        _blogUrlTextField = [NSTextField new];
        _blogUrlTextField.placeholderString = NSLocalizedString(@"Ghost Blog URL", nil);
        _blogUrlTextField.focusRingType = NSFocusRingTypeNone;
        _blogUrlTextField.bezelStyle = NSTextFieldRoundedBezel;
        _blogUrlTextField.translatesAutoresizingMaskIntoConstraints = NO;
    }
    
    return _blogUrlTextField;
}

- (NSButton *)addBlogButton
{
    if (!_addBlogButton) {
        _addBlogButton = [[NSButton alloc] init];
        _addBlogButton.title = NSLocalizedString(@"Add Blog", nil);
        _addBlogButton.translatesAutoresizingMaskIntoConstraints = NO;
        
        [_addBlogButton setButtonType:NSMomentaryLightButton];
        [_addBlogButton setBezelStyle:NSRoundedBezelStyle];
        [_addBlogButton setTarget:self];
        [_addBlogButton setAction:@selector(addBlog:)];
    }
    
    return _addBlogButton;
}

#pragma mark - Actions

- (void)addBlog:(id)sender
{
    NSURL *originalUrl = [NSURL URLWithString:self.blogUrlTextField.stringValue];
    NSURL *filteredUrl = [NSURL URLWithString:[NSString stringWithFormat:@"%@://%@", originalUrl.scheme, originalUrl.host]];
    NSURLRequest *request = [NSURLRequest requestWithURL:filteredUrl];
    
    AFHTTPRequestOperation *operation = [[AFHTTPRequestOperation alloc] initWithRequest:request];
    [operation setCompletionBlockWithSuccess:^(AFHTTPRequestOperation *operation, id responseObject)
    {
        // Filter HTML to get needed information
        TFHpple *doc = [[TFHpple alloc] initWithHTMLData:responseObject];
        
        NSArray *titles = [doc searchWithXPathQuery:@"//title"];
        NSArray *generators = [doc searchWithXPathQuery:@"//meta[@name='generator']"];
        NSString *adminUrlString = [NSString stringWithFormat:@"%@/ghost", filteredUrl.absoluteString];
        
        if (titles.count > 0 && generators.count > 0)
        {
            TFHppleElement *title = [titles objectAtIndex:0];
            TFHppleElement *generator = [generators objectAtIndex:0];
            
            if ([[generator objectForKey:@"content"] containsString:@"Ghost"])
            {
                // Don't add already added blogs
                for (NSDictionary *blog in [Utils blogs]) {
                    if ([blog[kBlogUrl] isEqualToString:adminUrlString]) return;
                }
                
                NSDictionary *blogInfo = @{kBlogName: title.text,
                                           kBlogUrl: adminUrlString};
                
                NSMutableArray *blogs = [NSMutableArray arrayWithArray:[Utils blogs]];
                [blogs insertObject:blogInfo atIndex:blogs.count];
                
                [[Utils userDefaults] setObject:blogs forKey:kBlogs];
            }
        }
        
    } failure:^(AFHTTPRequestOperation *operation, NSError *error) {
        NSLog(@"Error: %@", error);
    }];
    
    [operation start];
}

#pragma mark - Constraints

- (void)setUpConstraintsForViews:(NSDictionary *)views
{
    [self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:[logoImageView]" options:0 metrics:nil views:views]];
    [self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|-20-[blogUrlTextField]-20-|" options:0 metrics:nil views:views]];
    [self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|-20-[addBlogButton]-20-|" options:0 metrics:nil views:views]];
    [self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:|-80-[logoImageView]-60-[blogUrlTextField(30)]-5-[addBlogButton(30)]" options:0 metrics:nil views:views]];
    
    [self.view addConstraint:[NSLayoutConstraint constraintWithItem:self.logoImageView
                                                          attribute:NSLayoutAttributeCenterX
                                                          relatedBy:NSLayoutRelationEqual
                                                             toItem:self.view
                                                          attribute:NSLayoutAttributeCenterX
                                                         multiplier:1.0
                                                           constant:0.0]];
}

@end

File: Ghost/Controllers/ViewControllers/AddBlog/AddBlogViewController.swift
//
//  AddBlogViewController.swift
//  Ghost
//
//  Created by Enric Enrich on 19/09/15.
//  Copyright © 2015 Ghost. All rights reserved.
//

import Cocoa

class AddBlogViewController: NSViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.view.addSubview(self.logoImageView)
        self.view.addSubview(self.blogUrlTextField)
        self.view.addSubview(self.addBlogButton)
        
        let views: Dictionary<String, AnyObject> = ["logoImageView": self.logoImageView, "blogUrlTextField": self.blogUrlTextField, "addBlogButton": self.addBlogButton]
        
        self.setUpConstraints(views)
    }
    
    // Getters
    
    private lazy var logoImageView: NSImageView = {
        let imageView: NSImageView = NSImageView()
        imageView.image = NSImage(named: "GhostLogoAddBlog")
        imageView.translatesAutoresizingMaskIntoConstraints = false
        return imageView
    } ()
    
    private lazy var blogUrlTextField: NSTextField = {
        let textField: NSTextField = NSTextField()
        textField.placeholderString = NSLocalizedString("Ghost Blog Url", comment: "")
        textField.focusRingType = NSFocusRingType.None
        textField.bezelStyle = NSTextFieldBezelStyle.RoundedBezel
        textField.translatesAutoresizingMaskIntoConstraints = false
        return textField
    } ()
    
    private lazy var addBlogButton: NSButton = {
        let button: NSButton = NSButton()
        button.title = NSLocalizedString("Add Blog", comment: "")
        button.bezelStyle = NSBezelStyle.RoundedBezelStyle
        button.target = self
        button.action = "addBlog:"
        button.setButtonType(NSButtonType.MomentaryLightButton)
        button.translatesAutoresizingMaskIntoConstraints = false
        return button
    } ()
    
    // Actions
    
    func addBlog(sender: AnyObject?) {
        let originalUrl: NSURL = NSURL(string: self.blogUrlTextField.stringValue)!
        let filteredUrl: String = originalUrl.scheme + "://" + originalUrl.host!
        
        // Check if the "/ghost path works
        let manager: AFHTTPRequestOperationManager = AFHTTPRequestOperationManager()
        manager.responseSerializer = AFHTTPResponseSerializer()
        manager.GET(filteredUrl, parameters: nil, success: { (operation:AFHTTPRequestOperation!, responseObject:AnyObject!) -> Void in
            
            // Filter HTML to get needed information
            let doc: TFHpple = TFHpple(HTMLData: responseObject as! NSData)
            
            let titles: Array = doc.searchWithXPathQuery("//title")
            let generators: Array = doc.searchWithXPathQuery("//meta[@name='generator']")
            let adminUrlString: String = filteredUrl + "/ghost"
            
            if (titles.count > 0 && generators.count > 0)
            {
                let title: TFHppleElement = titles[0] as! TFHppleElement
                let generator: TFHppleElement = generators[0] as! TFHppleElement
                
                if (generator.objectForKey("content").containsString("Ghost"))
                {
                    // Don't add already added blogs
                    for blog: BlogViewModel in Utils.blogs() {
                        if (blog.urlString() == adminUrlString) { return }
                    }
                    
                    // Check if the "/ghost path works
                    let manager: AFHTTPRequestOperationManager = AFHTTPRequestOperationManager()
                    manager.responseSerializer = AFHTTPResponseSerializer()
                    manager.GET(adminUrlString, parameters: nil, success: { (operation:AFHTTPRequestOperation!, responseObject:AnyObject!) -> Void in
                        
                        if (operation.response!.statusCode == 200)
                        {
                            let blogInfo: Dictionary = [UserDefaultsBlogNameKey: title.text(), UserDefaultsBlogUrlKey: adminUrlString]
                            let blog: BlogViewModel = BlogViewModel(blogInfo: blogInfo)
                            
                            var blogs: Array = Utils.unarchivedBlogs()
                            blogs.append(NSKeyedArchiver.archivedDataWithRootObject(blog))
                            
                            Utils.userDefaults().setObject(blogs, forKey: UserDefaultsBlogsKey)
                            Utils.userDefaults().synchronize()
                            
                            NSNotificationCenter.defaultCenter().postNotificationName(DidAddBlogNotification, object: blog)
                        }
                        
                    }, failure: { (requestOperation, error) -> Void in
                            
                        print("Error")
                            
                    })
                }
            }
            
        }, failure: { (requestOperation, error) -> Void in
                
            print("Error")
                
        })
    }
    
    // Constraints
    
    private func setUpConstraints(views: Dictionary<String, AnyObject>) {
        self.view.addConstraints(NSLayoutConstraint.constraintsWithVisualFormat("H:[logoImageView]", options: NSLayoutFormatOptions.DirectionMask, metrics: nil, views: views))
        self.view.addConstraints(NSLayoutConstraint.constraintsWithVisualFormat("H:|-20-[blogUrlTextField]-20-|", options: NSLayoutFormatOptions.DirectionMask, metrics: nil, views: views))
        self.view.addConstraints(NSLayoutConstraint.constraintsWithVisualFormat("H:|-20-[addBlogButton]-20-|", options: NSLayoutFormatOptions.DirectionMask, metrics: nil, views: views))
        self.view.addConstraints(NSLayoutConstraint.constraintsWithVisualFormat("V:|-80-[logoImageView]-60-[blogUrlTextField(30)]-5-[addBlogButton(30)]", options: NSLayoutFormatOptions.DirectionMask, metrics: nil, views: views))
        
        self.view.addConstraint(NSLayoutConstraint(item: self.logoImageView, attribute: NSLayoutAttribute.CenterX, relatedBy: NSLayoutRelation.Equal, toItem: self.view, attribute: NSLayoutAttribute.CenterX, multiplier: 1.0, constant: 0.0))
    }
    
}
